custom/bank/keeper/keeper.go

```diff
@@ -66,13 +68,95 @@ func (k Keeper) SendCoins(ctx sdk.Context, fromAddr sdk.AccAddress, toAddr sdk.A
 func (k Keeper) SendCoinsFromModuleToModule(ctx sdk.Context, senderModule, recipientModule string, amt sdk.Coins) error {
 	senderAddr := k.ak.GetModuleAddress(senderModule)
 	if senderAddr == nil {
-		panic(sdkerrors.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", senderModule))
+		panic(errorsmod.Wrapf(customterratypes.ErrUnknownAddress, "senderModule address %s is nil", senderModule))
 	}
-
 	recipientAcc := k.ak.GetModuleAccount(ctx, recipientModule)
 	if recipientAcc == nil {
-		panic(sdkerrors.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", recipientModule))
+		panic(errorsmod.Wrapf(customterratypes.ErrUnknownAddress, "recipientModule address %s is nil", recipientModule))
 	}
 
 	return k.Keeper.SendCoins(ctx, senderAddr, recipientAcc.GetAddress(), amt)
 }
+
+// UndelegateCoins performs undelegation by crediting amt coins to an account with
+// address addr. For vesting accounts, undelegation amounts are tracked for both
+// vesting and vested coins. The coins are then transferred from a ModuleAccount
+// address to the delegator address. If any of the undelegation amounts are
+// negative, an error is returned.
+func (k Keeper) UndelegateCoins(ctx sdk.Context, moduleAccAddr, delegatorAddr sdk.AccAddress, amt sdk.Coins) error {
+	err := k.BlockBeforeSend(ctx, moduleAccAddr, delegatorAddr, amt)
+	if err != nil {
+		return err
+	}
+	k.TrackBeforeSend(ctx, moduleAccAddr, delegatorAddr, amt)
+
+	return k.Keeper.UndelegateCoins(ctx, moduleAccAddr, delegatorAddr, amt)
+}
+
+// DelegateCoins performs delegation by deducting amt coins from an account with
+// address addr. For vesting accounts, delegations amounts are tracked for both
+// vesting and vested coins. The coins are then transferred from the delegator
+// address to a ModuleAccount address. If any of the delegation amounts are negative,
+// an error is returned.
+func (k Keeper) DelegateCoins(ctx sdk.Context, delegatorAddr, moduleAccAddr sdk.AccAddress, amt sdk.Coins) error {
+	err := k.BlockBeforeSend(ctx, delegatorAddr, moduleAccAddr, amt)
+	if err != nil {
+		return err
+	}
+	k.TrackBeforeSend(ctx, delegatorAddr, moduleAccAddr, amt)
+
+	return k.Keeper.DelegateCoins(ctx, delegatorAddr, moduleAccAddr, amt)
+}
+
+// InputOutputCoins performs multi-send functionality. It accepts a series of
+// inputs that correspond to a series of outputs. It returns an error if the
+// inputs and outputs don't line up or if any single transfer of tokens fails.
+func (k Keeper) InputOutputCoins(ctx sdk.Context, inputs []banktypes.Input, outputs []banktypes.Output) error {
+	// ValidateInputsOutputs is executed in the k.Keeper.InputOutputCoins
+	// but we execute it here too just to avoid unnecessary computation
+	// if the inputs and outputs are invalid, that way we can assume things like
+	// the addreeses being valid and the inputs and outputs match balances
+	if err := banktypes.ValidateInputsOutputs(inputs, outputs); err != nil {
```

Comment:
don't think this is necessary. computation is paid by the tx sender so even if more computation is done and the transaction fails, the sender "pays for the mistake"